File: _CONTAINER/_BY_MONTH/_Feb/Thesis-physics-master/browser/engine/particleSystem.js
/**
 * @fileoverview Particle system.
 * @author sebastian.poreba@gmail.com (Sebastian Poreba)
 */

goog.provide('smash.ParticleSystem');


goog.require('smash.flags');



/**
 * @constructor
 */
smash.ParticleSystem = function() {
  /**
   * @type {!Array.<!smash.Particle>}
   */
  this.particles = [];

  /**
   * @type {!Array.<!smash.ParticleEmitter>}
   */
  this.emitters = [];

  if (smash.flags.DRAWING_ENABLED) {
    /**
     * @type {!Element}
     */
    this.canvas = window.document.createElement('canvas');
    this.canvas.width = smash.ParticleSystem.CANVAS_WIDTH;
    this.canvas.height = smash.ParticleSystem.CANVAS_HEIGHT;
    window.document.body.appendChild(this.canvas);

    /**
     * @type {!CanvasRenderingContext2D}
     */
    this.context = this.canvas.getContext('2d');

    /**
     * @type {!ImageData}
     */
    this.imageData = this.context.getImageData(0, 0,
        smash.ParticleSystem.CANVAS_WIDTH, smash.ParticleSystem.CANVAS_HEIGHT);

    /**
     * @type {!CanvasPixelArray}
     */
    this.pixels = this.imageData.data;
  }
};


/**
 * @const {number}
 */
smash.ParticleSystem.CANVAS_WIDTH = 1200;


/**
 * @const {number}
 */
smash.ParticleSystem.CANVAS_HEIGHT = 400;


/**
 *
 */
smash.ParticleSystem.prototype.step = function() {
  if (smash.flags.DRAWING_ENABLED) {
    for (var i = 0; i < smash.ParticleSystem.CANVAS_WIDTH *
        smash.ParticleSystem.CANVAS_HEIGHT * 4; i += 4) {
      this.pixels[i] = 0;
      this.pixels[i + 1] = 0;
      this.pixels[i + 2] = 0;
      this.pixels[i + 3] = 0;
    }
  }

  this.emitters.forEach(function(emitter) {
    this.particles.push.apply(this.particles,
        emitter.getNewParticles());
  }, this);

  var newParticles = [];
  this.particles.forEach(function(p) {
    p.step();
    if (p.positionX >= 0 &&
        p.positionX < smash.ParticleSystem.CANVAS_WIDTH &&
        p.positionY >= 0 &&
        p.positionY < smash.ParticleSystem.CANVAS_HEIGHT &&
        p.age < p.lifespan) {
      newParticles.push(p);
    }

    if (smash.flags.DRAWING_ENABLED) {
      var baseIndex =
          (Math.round(p.positionY) *
              smash.ParticleSystem.CANVAS_WIDTH +
              Math.round(p.positionX)) * 4;
      this.pixels[baseIndex] = 255;
      this.pixels[baseIndex + 1] = 0;
      this.pixels[baseIndex + 2] = 0;
      this.pixels[baseIndex + 3] = 255;
    }
  }, this);
  if (smash.flags.DRAWING_ENABLED) {
    this.context.putImageData(this.imageData, 0, 0);
  }

  this.particles = newParticles;
};


/**
 * @param {!smash.ParticleEmitter} emitter
 */
smash.ParticleSystem.prototype.addEmitter = function(emitter) {
  this.emitters.push(emitter);
};

File: _BY_MONTH/_Feb/Thesis-physics-master/doc/particles/particleSystem.cpp
/**
 * @fileoverview Particle system.
 * @author sebastian.poreba@gmail.com (Sebastian Poreba)
 */
#include "particleSystem.h"

smash::ParticleSystem::ParticleSystem() {
  this->particles = new std::vector<smash::Particle*>;
  this->emitters = new std::vector<smash::ParticleEmitter*>;
};

smash::ParticleSystem::~ParticleSystem() {
  this->particles->erase(this->particles->begin(), this->particles->end());
  delete this->particles;
  this->emitters->erase(this->emitters->begin(), this->emitters->end());
  delete this->emitters;
};

void smash::ParticleSystem::step() {
  for (std::vector<smash::ParticleEmitter*>::iterator it =
      this->emitters->begin(); it != this->emitters->end(); it++) {
    std::vector<smash::Particle*>* particleFromEmitters =
      (*it)->getNewParticles();
    this->particles->insert(this->particles->end(),
      particleFromEmitters->begin(), particleFromEmitters->end());
    delete particleFromEmitters;
  }

  std::vector<smash::Particle*> newParticles;

  for (std::vector<smash::Particle*>::iterator it =
      this->particles->begin(); it != this->particles->end(); it++) {
    smash::Particle* p = *it;
    p->step();
    if (p->positionX >= 0 &&
        p->positionX < smash::ParticleSystem::CANVAS_WIDTH &&
        p->positionY >= 0 &&
        p->positionY < smash::ParticleSystem::CANVAS_HEIGHT &&
        p->age < p->lifespan) {
      newParticles.push_back(p);
    } else {
      delete p;
    }
  };
  this->particles->swap(newParticles);
  newParticles.clear();
};

/**
 * @param emitter
 */
void smash::ParticleSystem::addEmitter(smash::ParticleEmitter* emitter) {
  this->emitters->push_back(emitter);
};
